### PR TITLE
Feature/core cloud fs usage

### PR DESCRIPTION
**Pull request by fpohtmeh-github**
Despite back-end isn't stable, feature is completed and can be reviewed

### CODE REVIEW

File: include/CloudFileSystem.h [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+
+private:
+    ModifiableCloudFileHandler createParentFolderFromInfo(const CloudFsFolder &fsFolder, const CloudFileHandler &oldFolder) const;
+    ModifiableCloudFileHandler createFolderFromInfo(const CloudFsFolderInfo &info, const CloudFileId &parentId, const QString &localPath) const;
+    ModifiableCloudFileHandler createFileFromInfo(const CloudFsFileInfo &info, const CloudFileId &parentId, const QString &localPath) const;
+
+    QPointer<CoreMessenger> m_coreMessenger;
+    std::optional<CoreMessengerCloudFs> m_coreFs;
+    std::atomic<CloudFileRequestId> m_requestId = 0;
+    Messenger *m_messenger;
```
Comment (SergeySeroshtan) [inline include/CloudFileSystem.h:97]: Use QPointer too.

File: include/controllers/CloudFilesController.h [hunk 2]
```diff
@@ -71,31 +73,45 @@ class CloudFilesController : public QObject
...
 signals:
+    void updateCloudFiles(const CloudFilesUpdate &update);
+
+    void notificationCreated(const QString &notification, const bool error) const;
     void errorOccurred(const QString &errorText);
 
     void displayPathChanged(const QString &path);
     void isRootChanged(const bool isRoot);
-    void isFetchingChanged(const bool isFetching);
+    void isLoadingChanged(const bool isLoading);
```
Comment (SergeySeroshtan) [inline include/controllers/CloudFilesController.h:83]: Using `const <primitive type>` in a function declaration is useless and takes more space.
Instead use `const <primitive type>` only in a function definition.

File: include/controllers/CloudFilesController.h [hunk 3]
```diff
@@ -71,31 +73,45 @@ class CloudFilesController : public QObject
...
+    void decLoadingCounter();
+
+    void onDbListFetched(const CloudFileHandler &parentFolder, const ModifiableCloudFiles &cloudFiles);
+    void onCloudFilesFetched(const CloudFileRequestId requestId, const ModifiableCloudFileHandler &parentFolder, const ModifiableCloudFiles &cloudFiles);
+    void onUpdateCloudFiles(const CloudFilesUpdate &update);
 
-    void onCloudFilesFetched(const CloudFileHandler &folder, const ModifiableCloudFiles &cloudFiles);
-    void onUpdateCloudFile(const CloudFileUpdate &update);
+    static bool fileIdLess(const ModifiableCloudFileHandler &a, const ModifiableCloudFileHandler &b);
+    static bool filesAreEqual(const ModifiableCloudFileHandler &a, const ModifiableCloudFileHandler &b);
```
Comment (SergeySeroshtan) [inline include/controllers/CloudFilesController.h:102]: Use `lhs` and `rhs` semantic instead if `a` and `b`.

File: include/messenger/CloudFsFolderId.h [hunk 4]
```diff
@@ -50,14 +50,16 @@ class CloudFsFolderId {
 
     bool isValid() const noexcept;
 
+    static CloudFsFolderId root();
+
+    bool operator<(const CloudFsFolderId& id) const;
+    bool operator>(const CloudFsFolderId& id) const;
+    bool operator==(const CloudFsFolderId& id) const;
+    bool operator!=(const CloudFsFolderId& id) const;
```
Comment (SergeySeroshtan) [inline include/messenger/CloudFsFolderId.h:58]: Using external operator is preferable as it guarantees stronger encapsulation.

File: include/operations/CloudFileOperation.h [hunk 5]
```diff
@@ -35,27 +35,34 @@
...
 
 signals:
-    void cloudFileUpdate(const CloudFileUpdate &update);
+    void cloudFilesUpdate(const CloudFilesUpdate &update);
 
 private:
-    static quint64 m_counter;
-    UserId m_userId;
+    static qsizetype m_counter;
+    Messenger *m_messenger;
```
Comment (SergeySeroshtan) [inline include/operations/CloudFileOperation.h:65]: Use QPointer with QObject pointers .

File: include/operations/DownloadCloudFileOperation.h [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+    void run() override;
+
+private:
+    void cleanup() override;
+
+    QString tempFilePath() const;
+
+    void onDownloadInfoGot(const CloudFileRequestId requestId, const CloudFileHandler &file, const QUrl &url, const QByteArray &encryptionKey);
+    void onGetDownloadInfoErrorOccurred(const CloudFileRequestId requestId, const QString &errorText);
+    void onProgressChanged(const quint64 bytesLoaded, const quint64 bytesTotal);
```
Comment (SergeySeroshtan) [inline include/operations/DownloadCloudFileOperation.h:62]: Using `const <primitive type>` in a function declaration is useless and takes more space.
Instead use `const <primitive type>` only in a function definition.

File: src/messenger/CloudFsFileId.cpp [hunk 7]
```diff
@@ -52,18 +52,18 @@ bool Self::isValid() const noexcept {
...
-bool operator==(const Self& lhs, const Self& rhs) {
-    return QString(lhs) == QString(rhs);
+bool Self::operator==(const Self& id) const {
+    return QString(*this) == QString(id);
 }
 
-bool operator!=(const Self& lhs, const Self& rhs) {
-    return QString(lhs) != QString(rhs);
+bool Self::operator!=(const Self& id) const {
+    return QString(*this) != QString(id);
```
Comment (SergeySeroshtan) [inline src/messenger/CloudFsFileId.cpp:68]: What is the motivation for such changes?

**Review by SergeySeroshtan — CHANGES_REQUESTED** [review]
Also as recommendation, do not put default initialisation values for properties with primitive types. It is redundant, except enumerations.
For instance
```cpp
struct A {
    int b = 0; // <-- this is redundant.
};

**Comment by fpohtmeh-github** [comment]
Fixed, please proceed

**Review by SergeySeroshtan — APPROVED** [review]

**Pull request merged**